File: src/main/java/com/shop/shopnearme/model/Shop.java
package com.shop.shopnearme.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import java.time.LocalDateTime;

@Entity
@Table(name = "shops")
public class Shop {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank
    private String name;

    @Column(length = 1000)
    private String description;

    @NotBlank
    private String address;

    @NotBlank
    private String city;

    @NotNull
    private Double latitude;

    @NotNull
    private Double longitude;

    private String phone;

    @JsonProperty("open")
    private boolean isOpen;

    @ManyToOne
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @ManyToOne
    @JoinColumn(name = "owner_id")
    private User owner;

    @Column(columnDefinition = "LONGTEXT")
    private String imageUrl;

    private String openingTime; // e.g., "09:00 AM"
    private String closingTime; // e.g., "09:00 PM"
    private boolean isClosedOnSunday;

    private String offers; // e.g., "10% off"
    private String services; // e.g., "Delivery, Pickup"
    private String tags; // e.g., "Grocery, Fresh"

    private boolean isVisible = true;
    private int views = 0;

    @Enumerated(EnumType.STRING)
    private CommunicationMode communicationMode = CommunicationMode.BOTH;

    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    // Getters and Setters

    // Getters and Setters

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getAddress() {
        return address;
    }

    public void setAddress(String address) {
        this.address = address;
    }

    public String getCity() {
        return city;
    }

    public void setCity(String city) {
        this.city = city;
    }

    public Double getLatitude() {
        return latitude;
    }

    public void setLatitude(Double latitude) {
        this.latitude = latitude;
    }

    public Double getLongitude() {
        return longitude;
    }

    public void setLongitude(Double longitude) {
        this.longitude = longitude;
    }

    public String getPhone() {
        return phone;
    }

    public void setPhone(String phone) {
        this.phone = phone;
    }

    @JsonProperty("manualOpen")
    public boolean isManualOpen() {
        return isOpen;
    }

    @JsonProperty("open")
    public boolean isOpen() {
        // Master Manual Toggle: If shopkeeper manually closed the shop, return false
        // immediately.
        if (!isOpen) {
            return false;
        }

        // Dynamic calculation based on time
        java.time.LocalDate today = java.time.LocalDate.now();
        java.time.LocalTime now = java.time.LocalTime.now();

        if (isClosedOnSunday && today.getDayOfWeek() == java.time.DayOfWeek.SUNDAY) {
            return false;
        }

        if (openingTime == null || closingTime == null || openingTime.isBlank() || closingTime.isBlank()) {
            // Fallback: if no time set, respect the manual field state (which we know is
            // true at this point)
            return true;
        }

        try {
            // Normalized patterns to try
            java.time.format.DateTimeFormatter formatter = java.time.format.DateTimeFormatter.ofPattern("hh:mm a",
                    java.util.Locale.ENGLISH);

            // Handle simple "HH:mm" case just in case
            if (!openingTime.toUpperCase().contains("AM") && !openingTime.toUpperCase().contains("PM")) {
                formatter = java.time.format.DateTimeFormatter.ofPattern("HH:mm");
            }

            java.time.LocalTime open = java.time.LocalTime.parse(openingTime, formatter);
            java.time.LocalTime close = java.time.LocalTime.parse(closingTime, formatter);

            // Handle overnight hours (e.g. 10 PM to 2 AM)
            if (open.isAfter(close)) {
                return now.isAfter(open) || now.isBefore(close);
            } else {
                return now.isAfter(open) && now.isBefore(close);
            }
        } catch (Exception e) {
            // In case of parse error, fallback to true since manual toggle is ON
            return true;
        }
    }

    public void setOpen(boolean open) {
        isOpen = open;
    }

    public Category getCategory() {
        return category;
    }

    public void setCategory(Category category) {
        this.category = category;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }

    public User getOwner() {
        return owner;
    }

    public void setOwner(User owner) {
        this.owner = owner;
    }

    public String getImageUrl() {
        return imageUrl;
    }

    public void setImageUrl(String imageUrl) {
        this.imageUrl = imageUrl;
    }

    public String getOpeningTime() {
        return openingTime;
    }

    public void setOpeningTime(String openingTime) {
        this.openingTime = openingTime;
    }

    public String getClosingTime() {
        return closingTime;
    }

    public void setClosingTime(String closingTime) {
        this.closingTime = closingTime;
    }

    public boolean isClosedOnSunday() {
        return isClosedOnSunday;
    }

    public void setClosedOnSunday(boolean closedOnSunday) {
        isClosedOnSunday = closedOnSunday;
    }

    public String getOffers() {
        return offers;
    }

    public void setOffers(String offers) {
        this.offers = offers;
    }

    public String getServices() {
        return services;
    }

    public void setServices(String services) {
        this.services = services;
    }

    public String getTags() {
        return tags;
    }

    public void setTags(String tags) {
        this.tags = tags;
    }

    public boolean isVisible() {
        return isVisible;
    }

    public void setVisible(boolean visible) {
        isVisible = visible;
    }

    public int getViews() {
        return views;
    }

    public void setViews(int views) {
        this.views = views;
    }

    public CommunicationMode getCommunicationMode() {
        return communicationMode;
    }

    public void setCommunicationMode(CommunicationMode communicationMode) {
        this.communicationMode = communicationMode;
    }
}

File: frontend_app/lib/models/shop.dart
import 'user.dart';
import 'category.dart';

enum CommunicationMode { WHATSAPP, CALL, BOTH }

class Shop {
  final int? id;
  final String name;
  final String? description;
  final String address;
  final String city;
  final double latitude;
  final double longitude;
  final String? phone;
  final bool isOpen;
  final bool manualOpen;
  final Category category;
  final User? owner;
  final String? imageUrl;
  final String? openingTime;
  final String? closingTime;
  final bool isClosedOnSunday;
  final String? offers;
  final String? services;
  final String? tags;
  final CommunicationMode communicationMode;
  double? distance;

  Shop({
    this.id,
    required this.name,
    this.description,
    required this.address,
    required this.city,
    required this.latitude,
    required this.longitude,
    this.phone,
    required this.isOpen,
    required this.manualOpen,
    required this.category,
    this.owner,
    this.imageUrl,
    this.openingTime,
    this.closingTime,
    this.isClosedOnSunday = false,
    this.offers,
    this.services,
    this.tags,
    this.communicationMode = CommunicationMode.BOTH,
  });

  Shop copyWith({
    int? id,
    String? name,
    String? description,
    String? address,
    String? city,
    double? latitude,
    double? longitude,
    String? phone,
    bool? isOpen,
    bool? manualOpen,
    Category? category,
    User? owner,
    String? imageUrl,
    String? openingTime,
    String? closingTime,
    bool? isClosedOnSunday,
    String? offers,
    String? services,
    String? tags,
    CommunicationMode? communicationMode,
    double? distance,
  }) {
    return Shop(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      address: address ?? this.address,
      city: city ?? this.city,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      phone: phone ?? this.phone,
      isOpen: isOpen ?? this.isOpen,
      manualOpen: manualOpen ?? this.manualOpen,
      category: category ?? this.category,
      owner: owner ?? this.owner,
      imageUrl: imageUrl ?? this.imageUrl,
      openingTime: openingTime ?? this.openingTime,
      closingTime: closingTime ?? this.closingTime,
      isClosedOnSunday: isClosedOnSunday ?? this.isClosedOnSunday,
      offers: offers ?? this.offers,
      services: services ?? this.services,
      tags: tags ?? this.tags,
      communicationMode: communicationMode ?? this.communicationMode,
    )..distance = distance ?? this.distance;
  }

  factory Shop.fromJson(Map<String, dynamic> json) {
    return Shop(
      id: json['id'],
      name: json['name'] ?? '',
      description: json['description'],
      address: json['address'] ?? '',
      city: json['city'] ?? '',
      latitude: (json['latitude'] ?? 0.0).toDouble(),
      longitude: (json['longitude'] ?? 0.0).toDouble(),
      phone: json['phone'],
      isOpen: json['open'] ?? false,
      manualOpen: json['manualOpen'] ?? json['open'] ?? false,
      category: Category.fromJson(json['category'] ?? {}),
      owner: json['owner'] != null ? User.fromJson(json['owner']) : null,
      imageUrl: json['imageUrl'],
      openingTime: json['openingTime'],
      closingTime: json['closingTime'],
      isClosedOnSunday: json['isClosedOnSunday'] ?? false,
      offers: json['offers'],
      services: json['services'],
      tags: json['tags'],
      communicationMode: CommunicationMode.values.firstWhere(
        (e) => e.name == json['communicationMode'],
        orElse: () => CommunicationMode.BOTH,
      ),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'address': address,
      'city': city,
      'latitude': latitude,
      'longitude': longitude,
      'phone': phone,
      'open': isOpen,
      'manualOpen': manualOpen,
      'category': category.toJson(),
      'owner': owner?.toJson(),
      'imageUrl': imageUrl,
      'openingTime': openingTime,
      'closingTime': closingTime,
      'isClosedOnSunday': isClosedOnSunday,
      'offers': offers,
      'services': services,
      'tags': tags,
      'communicationMode': communicationMode.name,
    };
  }
}
